components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1044,468 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
 
     return num_matches;
 }
+
+vector<QueryLogtype>
+Grep::generate_query_substring_logtypes(string& processed_search_string, ByteLexer& lexer) {
+    // Store substring logtypes in a set to avoid duplicates
+    vector<set<QueryLogtype>> query_substr_logtypes(processed_search_string.size());
+
+    // We need to differentiate between literal '*'/'?' and wildcards
+    auto [is_greedy_wildcard, is_non_greedy_wildcard, is_escape]
+            = get_wildcard_and_escape_locations(processed_search_string);
```

Comment:
Since `is_greedy_wildcard`, `is_non_greedy_wildcard`, `is_escape`, and `processed_search_string` are all linked, I think we should create a class like `SearchString` that encapsulates them all.

We can also create a method in that class that returns a `SearchStringView` which is a view of part of the search string. I think this would make a bunch of the indexing logic in these functions simpler.